django/db/models/sql/query.py

```diff
@@ -1373,10 +1376,12 @@ def add_q(self, q_object):
         # rel_a doesn't produce any rows, then the whole condition must fail.
         # So, demotion is OK.
         existing_inner = {a for a in self.alias_map if self.alias_map[a].join_type == INNER}
-        clause, _ = self._add_q(q_object, self.used_aliases)
+        clause, new_inner = self._add_q(q_object, self.used_aliases)
         if clause:
             self.where.add(clause, AND)
         self.demote_joins(existing_inner)
+        if promote_new_joins:
+            self.promote_joins(new_inner)
```

Comment:
I have a feeling something else if off here. The outer query's joining strategy should not have to special case inner queries as they are self contained expressions.

My guess is that something is getting mixed up wrt to aliases because the same model is being involved in the outer and inner queries.

Reply:
I hear that. I did find the following to be analogous and somewhat reassuring, though:
https://github.com/django/django/blob/4e4db426c5b0a18d1fd4dac6fc1342ab7762b288/django/db/models/query_utils.py#L100-L108